Beautify development process (#1443)
HULKs/hulk: docs/workflow/development.md

Author: No it doesn't, I only moved the reminder from `overview.md` to `development.md` in #1442 and didn't change anything else. The other changes are only prettier 🥴

Reviewer: Welp, it has conflicts now... 😅

Author: 😢

We use GitHub for our development workflow. This includes:

-   **Issues** for bug reports, feature requests, and general discussions
-   **Pull requests** for code reviews and contributions
-   **Projects** for organizing our development workflow
-   **Actions** for continuous integration (CI)

## The Development Project

At HULKs we use a quite heavily modified version of the [KanBan workflow](<https://en.wikipedia.org/wiki/Kanban_(development)>) (1).
The [_Current Board_](https://github.com/orgs/HULKs/projects/2/views/2) visualizes our current development status.
The [_List_](https://github.com/orgs/HULKs/projects/2/views/1) is mainly for organizing multiple iterations for a more long-term overview (important for dev-leads).
{ .annotate }

1. The modifications were mainly done to address the fact that we do not have fixed working hours.
   Limiting the amount of cards that are in progress does not work.
   Especially for members that are not present on a daily basis.

### Terms

> **Assignee:** The person that is responsible for a card (has the **exclusive** right to move a card(!) as soon as there is one).

Reviewer: ```suggestion
**Assignee:** The person that is responsible for a card; has the **exclusive** right to move a card as soon as there is one.
```

The assignee's job depends on the Kanban column:

-   No assignee: _Open_, _Done_
-   Assignee works on issue/pull request (if more or less progress happens on it): _In Progress_
-   Assignee is responsible to bring a pull request to `main` (e.g. reviewer, tester): _Request for Review_

!!! note

    Github distinguishes between _reviewers_ and _assignees_.
    We also do that.
    While the _assignee_ (in most cases one person) is responsible for the card and bring the branch into the main, the _reviewer(s)_ (can be more than one) can review the code at any time in addition to the notes that were given by the _assignee_.

>     In general: The more reviewers the better the resulting code (That being said: feel free to review code).

Reviewer: ```suggestion
    In general: More reviewers results in better code, so feel free to review anything you like.
```

**Author:** The person that created this card (issue/pull request).
Responsible for answering questions (issues/pull requests) and implement/discuss requested changes (pull requests).

### Columns

Our project board is divided into the following four columns:

| _Open_   | _In Progress_            | _Request for Review_    | _Done_   |
| -------- | ------------------------ | ----------------------- | -------- |
| -------- | Developer is responsible | Reviewer is responsible | -------- |

??? question "The _hidden_ column"

    The _hidden_ column is a special column that is not visible on the project board.
    It contains all open issues, that are not in _Open_ and that currently should not be worked on.

>     This is to keep the project board clean and focused on the current iteration.

Reviewer: Does this need to be here? I don't think this is relevant for most people.

Author: That's why I its hidden using the `???`
I think it still might be interesting, to clarify why somebody would move an issue there

Reviewer: Ok, fair, but I don't like the intense color of that admonition box. Its easily the most attention drawing item on the page.

Author: Okay, I changed it to info, do you prefer it like that?

#### Open

The _Open_ section contains by the dev-leads selected **issues** that are important and of high priority for the current iteration.

!!! note

    If you want to work on issues that are not in _Open_:
    Feel free to do so, but please create and issue beforehand and assign yourself to it, so that others know that you are working on it.

> Move (or add) an issue into _In Progress_ **and** assign yourself when you started working on it.

Reviewer: ```suggestion
Move or add an issue to _In Progress_ **and** assign yourself when you start working on it.
```

#### In Progress

The _In Progress_ section contains:

-   **issues** that have an assignee, but no open pull request.
    As soon as there is an open pull request fixing this issue, the issue card should be replaced by this pull request card (removing the Issue from the Project but not closing it as it is not fixed in the main yet).

>     The issue must then be mentioned with the `fixes #Issue_NO` in the pull requests description.

Reviewer: ```suggestion
    The issue must then be mentioned with the [`fixes` keyword](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests#linking-a-pull-request-to-an-issue) in the pull requests description.
```
Then we can remove the note below.
As much as I like admonitions, I think this page is quite overladen with with them already.

Author: Okay :(

-   **pull requests** that are not ready for review yet and are currently wip.

!!! note

    `fixes #Issue_NO` is a keyword on github.

>     Github will automatically close the mentioned issue whenever the corresponding pull request was merged.

Reviewer: ```suggestion
    Github will automatically close the mentioned issue whenever the corresponding pull request is merged.
```

>     Do **not** close issues that have not being fixed in the main yet (even if there is a pull request for it)!

Reviewer: This is already mentioned in the text above,

Author: What do you mean? I didn't see it

Reviewer: In the sentence before the link to github docs.
I think you can remove this whole admonition now.

Author: Okay, It doesn't include the information about the automatic closing but that's also fine with me. Removed

Move a pull request into _Request for Review_ when you have finished your work **and** tested the pull request yourself on relevant platforms.
At this stage a pull requests' description should be finalized (fill out the template properly).

#### Request for Review

The _Request for Review_ section contains pull requests that are ready to be reviewed.

!!! warning "Attention"

    This section is a prioritized FIFO queue. Add new cards at the bottom.
    The head of development might decide to move it further up if the pull request is rather important.

!!! tip

    If your pull request is not ready for review yet but you already want to get some feedback, move it to _Request for Review_ and create it as a [_draft pull request_](https://github.blog/news-insights/product-news/introducing-draft-pull-requests/).

!!! tip

    Enable _auto-merge_ if your pull request is not a _draft pull request_

> Assign yourself if you want to review this pull request.

Reviewer: ```suggestion
Assign yourself if you want to review a pull request.
```
